### PR TITLE
#138 always round scales to nearest 1/8 to support fractional scaling and other recent compositor changes; use --log-threshold debug to see details

### LINKED ISSUE #138: Overlapping outputs on some fractional scales
With some fractional scales, way-displays creates output layouts that overlap. For instance, using two 3840x2160 displays, and with
```
ARRANGE: ROW
ALIGN: TOP
SCALING: true
AUTO_SCALE: false
SCALE:
  - NAME_DESC: '!LG Electronics LG ULTRAFINE .*'
    SCALE: 1.8
```
a layout is created where the outputs have logical width 2133, but the second monitor is positioned at `x=2132`, according to `wayland-info`:
```
interface: 'zxdg_output_manager_v1',                     version:  3, name:  8
	xdg_output_v1
		output: 58
		name: 'DP-1'
		description: 'LG Electronics LG ULTRAFINE ... (DP-1)'
		logical_x: 2132, logical_y: 0
		logical_width: 2133, logical_height: 1200
	xdg_output_v1
		output: 57
		name: 'DP-2'
		description: 'LG Electronics LG ULTRAFINE ... (DP-2)'
		logical_x: 0, logical_y: 0
		logical_width: 2133, logical_height: 1200
```

(I'm not sure, but it seems that this might also be preventing some Sway key bindings from working (switch to prev/next output), which is how I noticed. From observation, this works when the scale is set such that the layout is not overlapping. In my case, scales of 1.7 and 1.9 both work here. Positioning the output at `x=2133` using `wlr-randr` also resolves this.)

Setup:
- way-displays 1.9.0
- Sway from [current master](https://github.com/swaywm/sway/commit/5c99b98805e8171d590d466b2eb21402aaf8de6b)

The compositor supports the new fractional scaling protocol, maybe that's the issue here? Afaik, this uses scales in increments of 1/120, while `wl_fixed_from_double()` which is used in computing the desired scale uses a base of 256.

It's even worse with scale 2.01, where `logical_x: 1909` (as set by way-displays) and `logical_width: 1912`. On the other hand, `120 * 2.01 = 241.2` and `3840 * 120/241 = 1912.03`, so that's probably where `logical_width` comes from.

Using scales that are multiples of 0.125 was fine in my tests (`logical_x` and `logical_width` matched exactly), which would make sense since 120 and 256 are both divisible by 8.

### PR DESCRIPTION
**Pull request by fberg**
fixes #138 

Attempts to fix #138 by using a different "scaling base" when the compositor supports the new [fractional-scale-v1](https://wayland.app/protocols/fractional-scale-v1) protocol.

With this PR, I observed no gaps or any overlap between the two outputs (each 3840x2160 pixels) at various tested scales. The issue with "flipping" scales (see discussion in #138) is also gone.

Tested on the following compositors:
- Sway 1.8.1
- Sway master (has fractional-scale-v1)
- labwc 0.6.6

I wanted to test on Hyprland 0.32.3 and Wayfire master as well (they would have fractional-scale-v1), but way-displays (even the released 0.9.0) fails to setup the layout when a fractional scale is used, so there is probably something else going on.

### CODE REVIEW

**Comment by fberg** [comment]
@alex-courtis I've included the commits from your branch (`138-fix-overlapping-outputs`, #144) in this PR since the behavior without fractional-scale-v1 is the same as in 53fef6873cd58df1abf081a163ba385f958d9120, so the updated tests were needed. Hope that's okay.

**Comment by fberg** [comment]
>I wanted to test on Hyprland 0.32.3 and Wayfire master as well

The issues there were probably due to my graphics driver or something, it works again now.

However, it appears that both Hyprland and Wayfire behave different than Sway here: the output sizes after configuration are the same as computed with this PR's code but _without_ the extra fractional-scale-v1 logic. So, perhaps this is a Sway bug after all and #144 would be sufficient? :confused:  Will have to investigate more.

**Comment by alex-courtis** [comment]
> However, it appears that both Hyprland and Wayfire behave different than Sway here: the output sizes after configuration are the same as computed with this PR's code but _without_ the extra fractional-scale-v1 logic. So, perhaps this is a Sway bug after all and #144 would be sufficient? 😕 Will have to investigate more.

#144 was just a hope. It covered river but not sway. We need a complete solution - this is the right direction.

I think we'll need to listen for the actual sizes after the scale is set. I'll branch an experiment off this branch.

File: inc/head.h [hunk 1]
```diff
@@ -26,6 +27,8 @@ struct HeadState {
 
 struct Head {
 
+	struct Displ *displ;
```
Comment (alex-courtis) [inline inc/head.h:30]: Once we are happy we can probably refactor this away. I'm not a fan of back references ;)

**Review by alex-courtis — COMMENTED** [review]
This is looking great! Many thanks for your speedy work...

**Comment by alex-courtis** [comment]
I've created a branch https://github.com/alex-courtis/way-displays/tree/138-retrieve-output-dimensions which adds xdg-output-unstable-v1 and listens to output events.

It's not hooked up to anything yet, however it is very useful for debugging, as it shows the actual logical x/y/w/h

File: inc/head.h (hunk 1 above)
Comment (fberg) [reply inc/head.h]: Removed this in d032ac6fe6243652da073c5f4c1de3316b9d6788 in favor of a `scaling_base` integer in `Head`.

**Comment by fberg** [comment]
>I've created a branch which adds xdg-output-unstable-v1 and listens to output events.

Great, that'll be very useful. It seems to me that taking an output's size after it was configured (and configuring them one after the other) is the only real way forward here, seeing as my changes don't work on some compositors.

Of course, one could still clamp the scale to multiples of 0.125 as an intermediate fix.

**Comment by alex-courtis** [comment]
> > I've created a branch which adds xdg-output-unstable-v1 and listens to output events.
> 
> Great, that'll be very useful. It seems to me that taking an output's size after it was configured (and configuring them one after the other) is the only real way forward here, seeing as my changes don't work on some compositors.
> 
> Of course, one could still clamp the scale to multiples of 0.125 as an intermediate fix.

This PR is a bugfix that should resolve the specific issues with fractional scaling compositors and doesn't affect existing functionality.

Clamping scales can be considered a feature (#40) and it doesn't seem to be needed to resolve this case.

Listening to xdg-output may not ever be necessary and adds a lot of complexity i.e. an extra set step.

Providing this branch passes all the test cases as per https://github.com/alex-courtis/way-displays/pull/145#issuecomment-1837655119 we should just merge this and release as a PATCH.

**Comment by alex-courtis** [comment]
Apologies for the sporadic replies; I should have mentioned that I'm only able to work on this project <1 day per week.

**Comment by alex-courtis** [comment]
**Alternative:** just clamp scales by implementing #40 

Do you think that would resolve this entire fractional scaling compositor issue?

It would still be useful to have the ability to determine if the compositor does fractional scaling, as it's likely that we will need that some time in the future.

**Comment by fberg** [comment]
>This PR is a bugfix that should resolve the specific issues with fractional scaling compositors and doesn't affect existing functionality.

While it fixes a bug for users of an unreleased Sway version, it does introduce the same problems for Hyprland though.
So I unfortunately don't see it as a clear improvement, but I'd be okay with merging it if you think this should be done.

>Do you think that [scale clamping] would resolve this entire fractional scaling compositor issue?

Yes, restricting allowed scales to multiples of 1/8ths should resolve all issues that I encountered in my tests.

**Comment by alex-courtis** [comment]
> Yes, restricting allowed scales to multiples of 1/8ths should resolve all issues that I encountered in my tests.

It seems that is the best solution here.

It's disappointing; I like the "fractional scaling choice" and I have a feeling we'll need it in the future.

Options:
1. create a new PR with hardcoded 1/8ths
1. rework this PR with hardcoded 1/8ths and keep the fractional-scale-v1 detection
1. create a new PR to address #40 with a minimum of 1/8

**Comment by fberg** [comment]
Sorry for my infrequent responses. I'm currently travelling.

I'd be happy to rework my MR to implement option 2, but will only be able to do so by end of January at the earliest. The other options also sound good to me though, and I wouldn't be mad if my PR ends up not being merged. :)

**Comment by alex-courtis** [comment]
> Sorry for my infrequent responses. I'm currently travelling.
> 
> I'd be happy to rework my PR to implement option 2, but will only be able to do so by end of January at the earliest. The other options also sound good to me though, and I wouldn't be mad if my PR ends up not being merged. :)

No worries at all. I'm in a similar situation.

**Comment by fberg** [comment]
Updated the PR to round the scale to a multiple of 1/8 if the monitor has fractional-scale-v1.

**Comment by alex-courtis** [comment]
> Updated the PR to round the scale to a multiple of 1/8 if the compositor has fractional-scale-v1.

I'm really happy with that: I've hardcoded fractional scaling on and the scales are sensible. It's working just fine with sway 1.8.1 and river master (fractional scaling). You've validated that it's resolved your initial failure case on sway master.

1. Less granular scales are desirable anyway: #40 , we would simply be adding an option to increase `HEAD_FRACTIONAL_SCALING_BASE`
2. Fractional scaling is coming; this change will be forced. We may as well "break" it now.
3. Non `HEAD_FRACTIONAL_SCALING_BASE` scales do produce a bad/blurry result when integer multiple scaling is used for apps that don't adopt fractional scaling.

∴let's just do this: remove the test for fractional scaling and always use the new codepaths.

**Comment by alex-courtis** [comment]
We should keep `have_fractional_scale_v1`. It should be logged at startup time to aid diagnosis of issues.

In a subsequest PR we can add xdg-output-unstable to retrieve the positions, but that will again be information / warning only.

**Comment by fberg** [comment]
Removed the check for the fractional scaling protocol, scales are now always rounded to the nearest multiple of 1/8th. Let me know if that's not what you intended.

>We should keep have_fractional_scale_v1. It should be logged at startup time to aid diagnosis of issues.

There's a debug log message in `listener_registry.c`.

**Comment by alex-courtis** [comment]
> Removed the check for the fractional scaling protocol, scales are now always rounded to the nearest multiple of 1/8th. Let me know if that's not what you intended.
> 
> > We should keep have_fractional_scale_v1. It should be logged at startup time to aid diagnosis of issues.
> 
> There's a debug log message in `listener_registry.c`.

Fantastic! I'll review / test this one on Monday.

**Review by alex-courtis — APPROVED** [review]
This is fantastic, the calculated scales for 1/120 1/256 always produce correct logical positions and dimensions.

I hope you don't mind, I enhanced the logging and added some tests.

**Comment by alex-courtis** [comment]
Real world testing:

```yaml
ARRANGE: ROW
ALIGN: BOTTOM
ORDER:
  - LG Display 0x05EF
  - M28U
  - Gigabyte M32Q
SCALING: TRUE
AUTO_SCALE: TRUE
SCALE:
  - NAME_DESC: LG Display 0x05EF
    SCALE: 2.3
  - NAME_DESC: M28U
    SCALE: 1.7
  - NAME_DESC: Gigabyte M32Q
    SCALE: 1.9
VRR_OFF:
  - LG Display 0x05EF
  - M28U
  - Gigabyte M32Q
```

released:
```
interface: 'zxdg_output_manager_v1',                     version:  3, name: 13
	xdg_output_v1
		output: 47
		name: 'HDMI-A-1'
		description: 'GIGA-BYTE TECHNOLOGY CO., LTD. M28U 21490B006081 (HDMI-A-1)'
		logical_x: 1113, logical_y: 0
		logical_width: 2259, logical_height: 1271
	xdg_output_v1
		output: 46
		name: 'eDP-1'
		description: 'LG Display 0x05EF (eDP-1)'
		logical_x: 0, logical_y: 645
		logical_width: 1112, logical_height: 625
	xdg_output_v1
		output: 45
		name: 'DP-6'
		description: 'GIGA-BYTE TECHNOLOGY CO., LTD. Gigabyte M32Q 0x0000064C (DP-6)'
		logical_x: 3373, logical_y: 512
		logical_width: 1348, logical_height: 758
interface: 'wl_output',                                  version:  4, name: 45
	name: DP-6
	description: GIGA-BYTE TECHNOLOGY CO., LTD. Gigabyte M32Q 0x0000064C (DP-6)
	x: 0, y: 0, scale: 2,
	physical_width: 700 mm, physical_height: 390 mm,
	make: 'GIGA-BYTE TECHNOLOGY CO., LTD.', model: 'Gigabyte M32Q',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 2560 px, height: 1440 px, refresh: 143.912 Hz,
		flags: current
interface: 'wl_output',                                  version:  4, name: 46
	name: eDP-1
	description: LG Display 0x05EF (eDP-1)
	x: 0, y: 0, scale: 3,
	physical_width: 310 mm, physical_height: 170 mm,
	make: 'LG Display', model: '0x05EF',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 2560 px, height: 1440 px, refresh: 59.998 Hz,
		flags: current
interface: 'wl_output',                                  version:  4, name: 47
	name: HDMI-A-1
	description: GIGA-BYTE TECHNOLOGY CO., LTD. M28U 21490B006081 (HDMI-A-1)
	x: 0, y: 0, scale: 2,
	physical_width: 630 mm, physical_height: 360 mm,
	make: 'GIGA-BYTE TECHNOLOGY CO., LTD.', model: 'M28U',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
		width: 3840 px, height: 2160 px, refresh: 30.000 Hz,
		flags: current
```

Logical positions and scales do not add up.


branch:
```
interface: 'zxdg_output_manager_v1',                     version:  3, name: 13
xdg_output_v1
output: 47
name: 'HDMI-A-1'
description: 'GIGA-BYTE TECHNOLOGY CO., LTD. M28U 21490B006081 (HDMI-A-1)'
logical_x: 1137, logical_y: 0
logical_width: 2194, logical_height: 1234
xdg_output_v1
output: 46
name: 'eDP-1'
description: 'LG Display 0x05EF (eDP-1)'
logical_x: 0, logical_y: 594
logical_width: 1137, logical_height: 640
xdg_output_v1
output: 45
name: 'DP-6'
description: 'GIGA-BYTE TECHNOLOGY CO., LTD. Gigabyte M32Q 0x0000064C (DP-6)'
logical_x: 3331, logical_y: 466
logical_width: 1365, logical_height: 768
interface: 'wl_output',                                  version:  4, name: 45
name: DP-6
description: GIGA-BYTE TECHNOLOGY CO., LTD. Gigabyte M32Q 0x0000064C (DP-6)
x: 0, y: 0, scale: 2,
	physical_width: 700 mm, physical_height: 390 mm,
	make: 'GIGA-BYTE TECHNOLOGY CO., LTD.', model: 'Gigabyte M32Q',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
width: 2560 px, height: 1440 px, refresh: 143.912 Hz,
	flags: current
	interface: 'wl_output',                                  version:  4, name: 46
	name: eDP-1
	description: LG Display 0x05EF (eDP-1)
	x: 0, y: 0, scale: 3,
	physical_width: 310 mm, physical_height: 170 mm,
	make: 'LG Display', model: '0x05EF',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
width: 2560 px, height: 1440 px, refresh: 59.998 Hz,
	flags: current
	interface: 'wl_output',                                  version:  4, name: 47
	name: HDMI-A-1
	description: GIGA-BYTE TECHNOLOGY CO., LTD. M28U 21490B006081 (HDMI-A-1)
	x: 0, y: 0, scale: 2,
	physical_width: 630 mm, physical_height: 360 mm,
	make: 'GIGA-BYTE TECHNOLOGY CO., LTD.', model: 'M28U',
	subpixel_orientation: unknown, output_transform: normal,
	mode:
width: 3840 px, height: 2160 px, refresh: 30.000 Hz,
	flags: current
```

Everything adds up.

**Pull request merged**